### PR TITLE
Issue/44 aw data annotations

### LINKED ISSUE #44: Validation for Lodging API Models
Add validation to fields on the Models in the Lodging API.

### LINKED ISSUE #52: LodgingRepo change
In \rvtrx-api-lodging\aspnet\RVTR.Lodging.DataContext\Repositories\LodgingRepo.cs the line where it says
rental.Status.Equals("available")
should instead be
rental.Status.ToLower().Equals("available")
in the database it uses "Available" rather than "available".

### PR DESCRIPTION
**Pull request by Jonny-Code**
resolves #44 and resolves #52 also fixes blocker for Campground needing reviews from lodgings objects

### CODE REVIEW

File: aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs [hunk 1]
```diff
@@ -22,6 +22,7 @@ public LodgingRepo(LodgingContext context) : base(context)
         .ThenInclude(ru => ru.Unit)
       .Include(l => l.Location)
         .ThenInclude(a => a.Address)
+      .Include(r => r.Reviews)
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs:25]: document on method, & also below

File: aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs [hunk 2]
```diff
@@ -63,7 +65,7 @@ public async Task<IEnumerable<LodgingModel>> LodgingByLocationAndOccupancy(int o
       {
         foreach (var rental in item.Rentals)
         {
-          if (rental.Status.Equals("available") && rental.Unit.Capacity >= occupancy && !filteredLodgings.Contains(item))
+          if (rental.Status.ToLower().Equals("available") && rental.Unit.Capacity >= occupancy && !filteredLodgings.Contains(item))
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs:68]: use case insensitive equals

File: aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs [hunk 3]
```diff
@@ -9,45 +9,60 @@ namespace RVTR.Lodging.ObjectModel.Models
...
+    [RegularExpression(@"^[A-Z]+[a-zA-Z""'\s-]*$", ErrorMessage = "Name must start with a capital letter and only use letters.")]
     public string City { get; set; }
 
     /// <summary>
-    ///
+    /// Country property must be US or USA
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Country is required")]
+    [RegularExpression(@"(^USA$)|(^US$)")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs:32]: error message

File: aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs [hunk 4]
```diff
@@ -9,45 +9,60 @@ namespace RVTR.Lodging.ObjectModel.Models
...
     public string StateProvince { get; set; }
 
     /// <summary>
-    ///
+    /// Street must be between 1 and 50 characters long
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Street is required")]
+    [MinLength(1)]
+    [MaxLength(50, ErrorMessage = "Street name is too long.")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs:65]: higher than 50

File: aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs [hunk 5]
```diff
@@ -9,33 +9,37 @@ namespace RVTR.Lodging.ObjectModel.Models
...
     /// <value></value>
     public AddressModel Address { get; set; }
 
     /// <summary>
-    ///
+    /// Latitude of the location model
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Latitude is required")]
+    [RegularExpression(@"^([0-8]{1}?[0-9]{1}(\.\d+)?|90(\.0+)?)[NS]$", ErrorMessage = "Latitude must be in the form 00.0000E")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs:34]: clean up regexes

File: aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs [hunk 6]
```diff
@@ -9,33 +9,37 @@ namespace RVTR.Lodging.ObjectModel.Models
...
     public string Latitude { get; set; }
 
     /// <summary>
-    ///
+    /// Longitude of the location model, validated to be in the format 00.000N, max of 180, min of 0
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Longitude is required")]
+    [RegularExpression(@"^(1([0-7]{1}[0-9]{1}(\.\d+)?|80(\.0+)?)|([0-9]{2}|[0-9]{1})(\.\d+))[EW]$", ErrorMessage = "Longitude must be in the form 00.0000N")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs:42]: EW vs NS

File: aspnet/RVTR.Lodging.ObjectModel/Models/LodgingModel.cs [hunk 7]
```diff
@@ -9,46 +9,50 @@ namespace RVTR.Lodging.ObjectModel.Models
...
     /// <value></value>
     public LocationModel Location { get; set; }
 
     /// <summary>
-    ///
+    /// Name of the lodging (required)
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Name is required")]
+    [MaxLength(20, ErrorMessage = "Max length is 20 characters")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/LodgingModel.cs:34]: more than 20

File: aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs [hunk 8]
```diff
@@ -10,14 +10,22 @@ public class RentalModel : IValidatableObject
   {
     public int Id { get; set; }
 
+    [Required(ErrorMessage = "Lot number can't be null.")]
+    [MaxLength(5, ErrorMessage = "Lot number must be 5 digits maximum")]
+    [RegularExpression(@"^[0-9]?[0-9]?[0-9]?[0-9]?[0-9]$")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs:15]: not just digits for this

File: aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs [hunk 9]
```diff
@@ -10,14 +10,22 @@ public class RentalModel : IValidatableObject
...
     public string LotNumber { get; set; }
 
     public RentalUnitModel Unit { get; set; }
 
+    [Required(ErrorMessage = "Status can't be null.")]
+    [RegularExpression(@"^([Bb]ooked|[Aa]vailable)$", ErrorMessage = "Status must be either 'Booked' or 'Available'")]
     public string Status { get; set; }
 
+    [Required(ErrorMessage = "Price can't be null.")]
+    [RegularExpression(@"^\d+(\.\d\d)?$", ErrorMessage = "Price must be in the form 00.00")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs:25]: not applicable, check for negative instead

**Review by escalonn — CHANGES_REQUESTED** [review]

File: aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs (hunk 1 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs:25]: Resolved by Alex

File: aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs (hunk 2 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.DataContext/Repositories/LodgingRepo.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs (hunk 3 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs (hunk 4 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/AddressModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs (hunk 5 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs (hunk 6 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/LocationModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs (hunk 8 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/LodgingModel.cs (hunk 7 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/LodgingModel.cs]: Resolved by Alex

File: aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs (hunk 9 above)
Comment (ddowd97) [reply aspnet/RVTR.Lodging.ObjectModel/Models/RentalModel.cs]: Resolved by Alex

**Review by markmoorerev — APPROVED** [review]

**Review by markmoorerev — APPROVED** [review]

File: aspnet/RVTR.Lodging.ObjectModel/Models/RentalUnitModel.cs [hunk 10]
```diff
@@ -15,33 +15,38 @@ public class RentalUnitModel : IValidatableObject
...
     /// <value></value>
     public RentalModel Rental { get; set; }
 
     /// <summary>
-    ///
+    /// the size of the rental unit (e.g. 5 x 5, 5x5, 5ft x 5ft, 5 yards x 5 yards etc.)
     /// </summary>
     /// <value></value>
+    [Required(ErrorMessage = "Size must exist")]
+    [RegularExpression(@"^\d+ ?([Ff]t|[Yy]ards|[Mm]eters|[Mm]) ?x ?\d+ ?([Ff]t|[Yy]ards|[Mm]eters|[Mm])$", ErrorMessage = "Size must be in the form '10 [unit?] x 10 [unit?]'")]
```
Comment (escalonn) [inline aspnet/RVTR.Lodging.ObjectModel/Models/RentalUnitModel.cs:49]: this is probably a bit over specific but fine

**Review by escalonn — APPROVED** [review]

**Pull request merged**